packages/ipfs-message-port-protocol/src/dag.js

```diff
@@ -0,0 +1,95 @@
+'use strict'
+
+const CID = require('cids')
+const { encodeCID, decodeCID } = require('./cid')
+
+/**
+ * @typedef {import('./data').JSONValue} JSONValue
+ */
+
+/**
+ * @template T
+ * @typedef {import('./data').StringEncoded<T>} StringEncoded
+ */
+
+/**
+ * @typedef {Object} EncodedCID
+ * @property {string} codec
+ * @property {Uint8Array} multihash
+ * @property {number} version
+ * @typedef {JSONValue} DAGNode
+ *
+ * @typedef {Object} EncodedDAGNode
+ * @property {DAGNode} dagNode
+ * @property {CID[]} cids
+ */
+
+/**
+ * @param {EncodedDAGNode} encodedNode
+ * @returns {DAGNode}
+ */
+const decodeNode = ({ dagNode, cids }) => {
+  // It is not ideal to have to mutate prototype chains like
+  // this, but it removes a need of traversing node first on client
+  // and now on server.
+  for (const cid of cids) {
+    decodeCID(cid)
+  }
+
+  return dagNode
+}
+
+exports.decodeNode = decodeNode
+
+/**
+ * Encodes DAG node for over the message channel transfer by collecting all
+ * the CID instances into an array so they could be turned back into CIDs
+ * without traversal on the other end.
+ *
+ * If `transfer` array is provided all the encountered `ArrayBuffer`s within
+ * this node will be added to transfer so they are moved across without copy.
+ * @param {DAGNode} dagNode
+ * @param {Transferable[]} [transfer]
+ * @returns {EncodedDAGNode}
+ */
+const encodeNode = (dagNode, transfer) => {
```

Comment:
I still think in the first instance it would have been much simpler to just use ipld on the UI thread and serialize the node to a ArrayBuffer and then transfer it with a codec that says how to interpret the data on the other end - in the same way we do with the http client.

It might not be as fast as a no-copy memory transfer but it would have made all of this a lot more straightforward.  Serialization speed is not where we've had performance bottlenecks in the past.

Reply:
Please note, this will become unnecessary once I'll get around to backporting this https://github.com/multiformats/js-multiformats/pull/23 into js-cids as per https://github.com/multiformats/js-cid/issues/111

If this is not a blocker, I would prefer to go with a current approach and address https://github.com/multiformats/js-cid/issues/111 to remove this encoding entirely instead of:

- Having to include codecs in both client and server bundles
- And going with a more sub-optimal approach.